### ISSUE TITLE
unneeded_concatenation_linter() can throw on valid usages

### ISSUE THREAD

**Issue by MichaelChirico** [r-lib/lintr#1344]
```
lint("c(matrix(1:10, 2, 5))\n", unneeded_concatenation_linter())
# <text>:1:1: style: Unneeded concatenation of a constant. Remove the "c" call.
# c(matrix(1:10, 1, 5))
# ^~~~~~~~~~~~~~~~~~~~~
```

`c()` can be used as here to concisely drop the dimensions on a matrix, so we can't just throw a lint on all `c(<expr>)` cases.

We have considered whether this usage should always be replaced by a more declarative version like `as.vector()` instead; perhaps we could do so with a linter parameter. In any case, this new lint marks a departure from 2.0.1 behavior.

Note that the new behavior _is_ good in some cases, e.g. it newly finds `c(1 - alpha / 2)` and `c(1:10)`, but I'm not sure there's a maintainable way to always catch these.

**Comment by AshesITR** [comment]
So we should customize the lint message in that case?
Maybe for all /SYMBOL_FUNCTION_CALL cases?

**Comment by MichaelChirico** [comment]
Yes, I think a custom message would be good. `SYMBOL` should be included too:

```
m <- matrix(1:10, 5, 2)
c(m)
```

**Comment by MichaelChirico** [comment]
Adding the node from `?c` here:

> `c` is sometimes used for its side effect of removing attributes except names, for example to turn an array into a vector. `as.vector` is a more intuitive way to do this, but also drops names.  Note that methods other than the default are not required to do this (and they will almost certainly preserve a class attribute).

I don't totally follow this, but the main takeaway is people may be using `c()` to accomplish name-preserving `as.vector()`. That doesn't seem like great practice, but reassures me that allowing `c(x)` by default is the better approach.

**Comment by MichaelChirico** [comment]
```r
x <- matrix(1:10, 5, 2)
c(x / 2)
```

I don't think we should lint that. I don't think this is a good example to refactor either -- needlessly verbose:

```
x <- matrix(1:10, 5, 2)
x_half <- x / 2
c(x_half)
```

I don't know immediately of any examples where `as.vector()` can't be used, but both `c()` and `as.vector()` are generics, so it's easy to imagine there are cases, e.g. a `c()` method is defined but `as.vector()` is not.

So for now, I think we should be conservative and not lint.

**Comment by MichaelChirico** [comment]
PS I did leave a carve-out for `a:b`, but actually, we shouldn't lint that either:

```
x <- factor(letters)
y <- factor(LETTERS)

x:y
```

**Comment by MichaelChirico** [comment]
Hid a comment about `c(x:y)` when `x` and `y` are factors that was mistaken. That case is the same as any other `c(factor)` usage.

**Comment by AshesITR** [comment]
> ```r
> x <- matrix(1:10, 5, 2)
> c(x / 2)
> ```
> 
> I don't think we should lint that. I don't think this is a good example to refactor either -- needlessly verbose:

`x <- (1:10) / 2` does the same, and `as.vector(x / 2)` returns the same, so I think it should be linted.
Still failing to find a valid piece of code that would require `c(symbol / 42)` to work (i.e. needs names) and can't be trivially refactored to something nicer.

**Comment by MichaelChirico** [comment]
but we don't always have control of generating `x`. e.g. `y <- other_pkg::foo(x); c(y / 2)`.

**Comment by AshesITR** [comment]
`y <- other_pkg::foo(x) / 2; c(y)`
I feel like we trade a rare false positive for many false negatives here.

**Comment by AshesITR** [comment]
Hmm, okay.

**Comment by MichaelChirico** [comment]
As a compromise I will emphasize that we encourage changing the default in the NEWS item. Also sets us up to change the default in the future if we get more confidence/feedback about it after release.
